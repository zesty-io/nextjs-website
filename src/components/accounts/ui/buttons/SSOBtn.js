import { Stack, Typography } from '@mui/material';
import { grey } from '@mui/material/colors';
import { accounts } from 'components/accounts/constants';
import React, { useEffect } from 'react';

const Index = ({
  image = accounts.logos.google,
  title = 'Sign in with Google',
  bodyColor = '#fff',
  logoColor = '#fff',
  textColor = '#333333',
  borderColor = grey[500],
  href,
}) => {
  const handleClick = () => {
    const popup = window.open(href, 'Pop up window');
    popup.postMessage('message', window.location.origin);
  };

  useEffect(() => {
    const redirectLogic = (message) => {
      if (message.origin === 'https://auth.api.zesty.io') {
        if (message.data.source === 'zesty' && message.data.status === '200') {
          let referrer = window.document.referrer;
          referrer =
            window.document.referrer === '' ||
            window.document.referrer.includes('/logout/') ||
            window.document.referrer.includes('/login/')
              ? window.location.origin + '/dashboard/'
              : window.document.referrer.replace(/^.*\/\/[^\/]+/, '');
          message.source.close();
          const sessionPrevUrl = sessionStorage.getItem('prevUrl');
          const prevUrl = window.document.referrer.replace(/^.*\/\/[^\/]+/, '');

          sessionStorage.removeItem('prevUrl');
          if (prevUrl === sessionPrevUrl || referrer.indexOf('/docs/') === 0) {
            window.location.href = referrer;
          } else {
            window.location.href = '/dashboard/';
          }
        }
      }
    };
    window.addEventListener('message', redirectLogic);
    return () => window.removeEventListener('message', redirectLogic);
  }, []);

  return (
    <Stack
      onClick={handleClick}
      direction={'row'}
      gap={1}
      alignItems="center"
      width={1}
      sx={{
        borderRadius: '3px',
        border: `1px solid ${borderColor}`,
        cursor: 'pointer',
        background: bodyColor,
        '&:hover': {
          filter: 'contrast(120%)',
          boxShadow: 2,
        },
      }}
    >
      <Stack
        p={1}
        bgcolor={logoColor}
        alignItems={'center'}
        justifyItems="center"
        sx={{
          borderRadius: '3px 0 0 3px',
        }}
      >
        <img src={image} alt={title} height={'24px'} width="24px" />
      </Stack>
      <Stack
        width={1}
        textAlign="center"
        sx={{
          color: textColor,
        }}
      >
        <Typography
          fontWeight={'500'}
          mr={6}
          color={textColor}
          whiteSpace={'nowrap'}
        >
          {title}
        </Typography>
      </Stack>
    </Stack>
  );
};
export const SSOBtn = React.memo(Index);
