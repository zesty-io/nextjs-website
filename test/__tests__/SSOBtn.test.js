import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen, fireEvent } from '@testing-library/react';
import { SSOBtn } from '../../src/components/accounts/ui/buttons/SSOBtn.js';

describe('SSO btn tests', () => {
  const title = 'Sign in with Google';
  const href = 'https://example.com';
  const image = 'path/to/image.png';
  const bodyColor = '#FFFFFF';
  const borderColor = '#CCCCCC';
  beforeEach(() => {
    render(
      <SSOBtn
        borderColor={borderColor}
        bodyColor={bodyColor}
        title={title}
        href={href}
        image={image}
      />,
    );
  });

  test('renders SSOBtn component', () => {
    const ssoBtnElement = screen.getByTestId('sso-btn');
    expect(ssoBtnElement).toBeInTheDocument();
  });
  test('renders title correctly', () => {
    const titleElement = screen.getByText(title);
    expect(titleElement).toBeInTheDocument();
  });
  test('displays image correctly', () => {
    const imageElement = screen.getByAltText('Sign in with Google');
    expect(imageElement).toHaveAttribute('src', image);
  });
  test('applies correct background color', () => {
    const ssoBtnElement = screen.getByTestId('sso-btn');
    expect(ssoBtnElement).toHaveStyle(`background: ${bodyColor}`);
  });
  test('applies correct border color', () => {
    const ssoBtnElement = screen.getByTestId('sso-btn');
    expect(ssoBtnElement).toHaveStyle(`border: 1px solid ${borderColor}`);
  });
  test('matches snapshot', () => {
    const tree = renderer
      .create(
        <SSOBtn
          borderColor="#CCCCCC"
          bodyColor="#FFFFFF"
          title="Sign in with Google"
          href="https://example.com"
          image="path/to/image.png"
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
