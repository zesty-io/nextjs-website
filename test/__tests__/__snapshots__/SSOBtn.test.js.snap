// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SSO btn tests matches snapshot 1`] = `
<div
  className="css-8736iw-MuiStack-root"
  data-testid="sso-btn"
  onClick={[Function]}
>
  <div
    className="css-1exl4mu-MuiStack-root"
  >
    <img
      alt="Sign in with Google"
      height="24px"
      src="path/to/image.png"
      width="24px"
    />
  </div>
  <div
    className="css-1b4q6aw-MuiStack-root"
  >
    <p
      className="MuiTypography-root MuiTypography-body1 css-1o8ifbq-MuiTypography-root"
    >
      Sign in with Google
    </p>
  </div>
</div>
`;
